update gtr doc (#198)
karmada-io/community: cncf_process/general_technical_review.md

Reviewer: ```suggestion
- [Survey](https://docs.google.com/document/d/1lOXHfpLiA0sg5dJr7ye9E0Qemd_YvOS-TACfsA3xQig/edit?usp=sharing) for installation and operation.
  - Join the [mailing list](https://groups.google.com/g/karmada/c/Xe489XICWqs/m/7i-sPn8cAAAJ) for access.
```

Minor link cleanup

Reviewer: This statement and the following contradict eachother slightly. If an upgrade fails for any reason - is rollback supported?

Author: I noticed that Helm supports rollback. I'll update the answer

Author: I have updated all the rollback-related answers.

# General Technical Review - Karmada / Incubation

- **Project:** Karmada
- **Project Version:** v0.17.0
- **Website:** https://karmada.io/
- **Date Updated:** 2026-04-30
- **Template Version:** v1.0

> - **Template Link:** [cncf/toc/general-technical-questions.md](https://github.com/cncf/toc/blob/42ee695dd281dd913400d242416c75010dc29444/.archive/resources/toc-supporting-guides/general-technical-questions.md)

Reviewer: ```suggestion
- **Template Link:** [cncf/toc/general-technical-questions.md](https://github.com/cncf/toc/blob/42ee695dd281dd913400d242416c75010dc29444/toc_subprojects/project-reviews-subproject/general-technical-questions.md)
```

  - Feature requests and proposals are collected from:
    - GitHub issues and proposals submitted by contributors
    - Feedback from users during community meetings

>     - Input from the [Karmada Adopter Group](https://github.com/karmada-io/community/tree/7782163b8e140716a44d18ac17ed8a4c48e8dd5b/adopter-group)

Reviewer 2: ```suggestion
    - Input from the [Karmada Adopter Group](https://github.com/karmada-io/community/tree/main/adopter-group)
```
Is there a specific reason you use the permanent link?

Author: > Tips:
> - You **CAN** link out to your own project's documentation, but be sure to link to it in a _versioned_ form. e.g. link to it at a specific commit instead of the `main` branch, or versioned website.

https://github.com/cncf/toc/blob/42ee695dd281dd913400d242416c75010dc29444/toc_subprojects/project-reviews-subproject/general-technical-questions.md?plain=1#L48

When linking to GitHub, I always try to use permanent links to ensure the documentation's accuracy and longevity

Reviewer 2: For the Karmada Adopter Group, it doesn't have to use the permanent link, as we want to show the latest information.

Author: Makes sense to me, updated

  - Simplify the process of deploying applications across multiple Kubernetes clusters.
  - Integrate Karmada into CI/CD pipelines for cross-cluster application deployment and management.
  - Ensure consistent application deployment, perform rolling updates, and manage application lifecycles across multiple clusters.

> - **AI Infrastructure Platform Teams / GPU Platform Operators**:

Reviewer 2: ```suggestion
- **AI Infrastructure Platform Teams:
```
The `GPU Platform Operators` are part of the infra teams.

>   - Use Karmada to build and operate multi-cluster AI infrastructure, especially for large-scale inference workloads.

Reviewer 2: ```suggestion
  - Use Karmada to build and operate multi-cluster AI infrastructure for large-scale training and inference workloads.
```
A lot of users are using Karmada for training.

- Resource status collection and aggregation  
- Global search for resources and events  
- Multi-cluster service discovery

> - AI/Big data workloads management across multiple clusters

Reviewer 2: This is now the primary use case. You can see that more and more users are using Karmada for AI workload.

Author: Agree, I'll move it to the primary use case section

Karmada’s feature lifecycle uses the Kubernetes model. Features are initially implemented using alpha version strings in the API and are feature gated. Features then graduate to beta and the feature gate is removed as part of the release lifecycle.

In addition, we provide conversion webhooks when CRD versions are updated. For example, webhook `resourcebindings.work.karmada.io` is used to update binding to v1alpha2.

## Day 2 - Day-to-Day Operations Phase

### Scalability/Reliability

#### Describe how the project increases the size or count of existing API objects.

Overall, the total object count grows approximately linearly with the number of workloads and target clusters.
When a user deploys a workload together with a PropagationPolicy in Karmada, the control plane creates a corresponding ResourceBinding object. It then creates one Work object for each target cluster selected for propagation. As a result, object growth is primarily fan-out based: a single source workload may produce one ResourceBinding and multiple per-cluster Work objects.

#### Describe how the project defines Service Level Objectives (SLOs) and Service Level Indicators (SLIs).

> Scalability and performance are key characteristics of multi-cluster federation. Karmada provides a large set of [measurable metrics](https://karmada.io/docs/next/reference/instrumentation/metrics). Based on these metrics, Karmada community defines the following [SLIs and SLOs](https://karmada.io/blog/2022/10/26/test-report#slisslos) to evaluate the service quality of multi-cluster federation.

Reviewer: `https://karmada.io/blog/2022/10/26/test-report/#slisslos` appears to be a good consolidation of what is being prepared below. We should weigh whether this is meaningful to include or not.

Author: Sorry, I don't get it. Could you elaborate on that a bit more?

Reviewer: no required change - This was to say - there was interesting and existing content that may have been useful for deriving the definition if SLOs and SLIs.

- Memory: Memory usage may increase because controllers and caches must hold more project-managed objects and more per-cluster state. For example, `karmada-search` is memory-intensive, as it mirrors the full state of registered resources from every member cluster. As a result, its memory cost grows multiplicatively with the number of clusters, resource types, and object counts.
- Storage: Storage usage may increase primarily in `karmada-etcd`, because the project creates additional objects.
- Throughput: The impact on the existing system is generally limited. The throughput increase mainly occurs within Karmada’s own control plane, where components such as `karmada-controller-manager` consume additional API read/write capacity from `karmada-apiserver`.

Overall, the increase is expected to be driven mainly by the number of workloads, policies, and target clusters, and to grow approximately linearly with those scaling dimensions.

#### Describe which conditions enabling / using this project would result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)

> During the process of enabling Karmada, the issue of "too many open files" may sometimes occur. Administrators can resolve this by using the following configurations:

Reviewer: More specifically this is from the webhook nodes and is covered in karmada troubleshooting documentation correct? May be worth a mention. 

[link](https://karmada.io/docs/troubleshooting/trouble-shooting/#karmada-webhook-keeps-on-crashing-due-to-too-many-open-files)

```bash
sysctl -w fs.inotify.max_user_watches=100000
sysctl -w fs.inotify.max_user_instances=100000
```

In addition, most Karmada components are CPU-intensive, so platform administrators need to ensure that the nodes have sufficient capacity to run Karmada components.

#### Describe the load testing that has been performed on the project and the results.

Karmada has performed both large-scale load testing and targeted performance testing.

> - [Test Report on Karmada's Support for 100 Large-Scale Clusters](https://karmada.io/blog/2022/10/26/test-report)

Reviewer: This is nice to see added as it was a point of interest when I was looking through the blog that I think should definitely be included in the review. No action needed.

    - [Overview of performance improvements for v1.15](https://github.com/karmada-io/karmada/issues/6516)
- Karmada also provides [CI-integrated](https://github.com/karmada-io/karmada/blob/cef28e8d92bbf8e4b209e066aed1e6e39ea09625/.github/workflows/ci-performance-compare.yaml) and [script-based](https://github.com/karmada-io/karmada/blob/cef28e8d92bbf8e4b209e066aed1e6e39ea09625/hack/performance/README.md) performance testing capabilities, making it easier to run repeatable load tests in both automated and local environments.

#### Describe the recommended limits of users, requests, system resources, etc. and how they were obtained.

Based on published [large-scale testing](https://karmada.io/blog/2022/10/26/test-report), Karmada is known to work at the scale of around 100 managed clusters. Scaling beyond that point may require increasing the resource requests/limits and replica counts of key Karmada control-plane components, as well as tuning supporting dependencies such as the Kubernetes API server and etcd.

#### Describe which resilience pattern the project uses and how, including the circuit breaker pattern.

> Karmada uses the Kubernetes resilience pattern to manage controllers with HA. If one replica fails, another part of the control plane takes responsibility to orchestrate workloads.

Reviewer: A little brief but not inaccurate. I've gathered there are a few mechanisms regarding reslience - from control plane operations (is there leader election?) - to the [HA documentation](https://karmada.io/docs/installation/ha-installation), to state durability with etcd. 

I did find a `Remedy` resource and I was unsure if it would be a good fit here or elsewhere as well.

### Observability Requirements

#### Describe the signals the project is using or producing, including logs, metrics, profiles and traces. Please include supported formats, recommended configurations and data storage.

Karmada produces several standard observability signals, including metrics, events, logs, and runtime profiles.

> - **Metrics**: Karmada exposes [Prometheus-compatible metrics](https://karmada.io/docs/next/reference/instrumentation/metrics) to report runtime status and component behavior. These metrics are intended to be scraped by Prometheus or another compatible monitoring backend.

Reviewer: solid references 👍 - no action required

#### Describe any dashboards the project uses or implements as well as any dashboard requirements.

Karmada provides and uses two main dashboard options:
- **Karmada Dashboard**: [Karmada Dashboard](https://github.com/karmada-io/dashboard) is a general-purpose, web-based control panel for Karmada, a multi-cluster management project. It provides a centralized UI for managing and observing Karmada resources and clusters.
- **Grafana Dashboards**: Karmada also provides [production-ready Grafana dashboards](https://karmada.io/docs/next/administrator/monitoring/karmada-observability/#grafana-dashboards) that can be downloaded and imported directly. These dashboards offer comprehensive observability coverage for Karmada components and operations.

**Requirements**:
Karmada Dashboard requires deployment of the separate dashboard project. The Grafana dashboards require a monitoring stack capable of collecting Karmada metrics, typically Prometheus for scraping and storage, and Grafana for visualization.

#### Describe how the project surfaces project resource requirements for adopters to monitor cloud and infrastructure costs, e.g. FinOps.

> Karmada does not currently implement a dedicated FinOps capability. Instead, it exposes standard operational and resource-related signals that adopters can use to monitor infrastructure footprint and estimate cloud costs. These include [configurable CPU and memory requests/limits](https://github.com/karmada-io/karmada/blob/cef28e8d92bbf8e4b209e066aed1e6e39ea09625/operator/pkg/apis/operator/v1alpha1/type.go#L679-L682) for Karmada components, [Prometheus-compatible metrics](https://karmada.io/docs/next/reference/instrumentation/metrics), logs, and [Grafana dashboards](https://karmada.io/docs/next/administrator/monitoring/karmada-observability/#grafana-dashboards). In practice, adopters typically combine these signals with external monitoring and cost-management systems to perform FinOps analysis.

Reviewer: It may be worth a mention that users _could_ use per-member-cluster allocation guages - such as [cluster_cpu_allocated_number](https://karmada.io/docs/administrator/monitoring/karmada-observability/#cluster_cpu_allocated_number) to correlate resource demands with member cluster spend.

Operators can use the following SLIs to assess the health and service quality of Karmada:

- **P99 latency of mutating API calls** (including POST, PUT, DELETE, and PATCH) to a single resource object over the last 5 minutes.
- **P99 latency of non-streaming read-only API calls** (including GET and LIST) over the last 5 minutes, measured separately by request scope.
- **P99 resource distribution latency**, measured from the time a resource template and policy are submitted to the federated control plane to the time the resource is created on the member cluster.
- **Cluster registration latency**, measured from the time a cluster is registered in the federation control plane to the time its status can be collected.
- **Resource usage**, measured as the amount of resources required for the federation control plane to operate normally after managing a given number of clusters.

These SLIs are backed by Karmada's measurable metrics and can be used by operators to monitor the runtime health, responsiveness, and scalability of the service.

### Dependencies

#### Describe the specific running services the project depends on in the cluster.

> Karmada’s core runtime dependency is Kubernetes itself. It does not require additional mandatory third-party in-cluster services beyond the Kubernetes APIs and its own Karmada control-plane components. Optional services such as Prometheus, Grafana, or Dashboard can be integrated for observability and UI, but they are not required for core functionality.

Reviewer: may consider elaborating on what from kubernetes is the project dependent on. Outside of the purpose it is providing.

Karmada’s feature lifecycle uses the Kubernetes model. Features are initially implemented using alpha version strings in the API and are feature gated. Features then graduate to beta and the feature gate is removed as part of the release lifecycle.

In addition, we provide conversion webhooks when CRD versions are updated. For example, webhook `resourcebindings.work.karmada.io` is used to update binding to v1alpha2.

## Day 2 - Day-to-Day Operations Phase

### Scalability/Reliability

#### Describe how the project increases the size or count of existing API objects.

Overall, the total object count grows approximately linearly with the number of workloads and target clusters.
When a user deploys a workload together with a PropagationPolicy in Karmada, the control plane creates a corresponding ResourceBinding object. It then creates one Work object for each target cluster selected for propagation. As a result, object growth is primarily fan-out based: a single source workload may produce one ResourceBinding and multiple per-cluster Work objects.

> #### Describe how the project defines Service Level Objectives (SLOs) and Service Level Indicators (SLIs).

Reviewer 2: Just a reminder, we are going to have a guide for users to set up their SLO. We can get back and make an update here once the PR https://github.com/karmada-io/website/pull/995 gets merged.

Author: and this one: https://github.com/karmada-io/karmada/issues/7385 for question "Describe the load testing that has been performed on the project and the results."

Reviewer 2: Sure.

Author: and this one: https://github.com/karmada-io/website/issues/1014 for question "Describe how the project is handling certificate rotation and mitigates any issues with certificates."

Based on published [large-scale testing](https://karmada.io/blog/2022/10/26/test-report), Karmada is known to work at the scale of around 100 managed clusters. Scaling beyond that point may require increasing the resource requests/limits and replica counts of key Karmada control-plane components, as well as tuning supporting dependencies such as the Kubernetes API server and etcd.

#### Describe which resilience pattern the project uses and how, including the circuit breaker pattern.

Karmada uses the Kubernetes resilience pattern, which includes the following key elements:

- **Control-plane HA:** Key control-plane components such as `karmada-controller-manager` can be deployed with multiple replicas and use leader election. If the active replica fails, another replica can take over and continue reconciliation.
- **[ETCD HA:](https://karmada.io/docs/next/installation/ha-installation/#options-for-highly-available-topology)** Karmada relies on etcd for control-plane state durability. Depending on the deployment model, adopters can use either Stacked etcd topology or External etcd topology to state durability with etcd.

> - **[Health-based scheduling isolation and failover:](https://karmada.io/docs/next/userguide/failover/failover-analysis)** When a member cluster is detected as unhealthy, Karmada can taint it with `NoSchedule` to stop new workloads from being placed there unless explicitly tolerated. If the unhealthy condition persists for a configurable period (5 minutes by default, controlled by `failover-eviction-timeout`), Karmada can further apply `NoExecute`, and workloads that exceed their toleration time can be evicted and failed over to healthy clusters.

Reviewer 2: ```suggestion
- **[Health-based scheduling isolation and failover:](https://karmada.io/docs/next/userguide/failover/failover-analysis)** Karmada uses taints on the Cluster object to control application scheduling and execution. The system automatically adds `NoSchedule` taints (e.g., `cluster.karmada.io/not-ready`, `cluster.karmada.io/unreachable`) based on cluster health conditions, preventing new workloads from being placed on unhealthy clusters. For more advanced scenarios, the `ClusterTaintPolicy` API allows administrators to define custom taint rules based on arbitrary cluster conditions. When the Failover feature gate is explicitly enabled, `NoExecute` taints can trigger eviction of workloads that do not tolerate the taint, enabling automatic failover to healthy clusters.
```
NOTE: the `failover-eviction-timeout` has been deprecated and has now been removed.
